content/docs/learn/typed-errors/validation.md

```diff
@@ -0,0 +1,312 @@
+-->
+
+```kotlin
+sealed interface BookValidationError
+object EmptyTitle: BookValidationError
+object NoAuthors: BookValidationError
+data class EmptyAuthor(val index: Int): BookValidationError
+
+data class Book private constructor(val title: String, val authors: List<Author>) {
+  companion object {
+    operator fun invoke(
+      title: String, authors: List<String>
+    ): Either<BookValidationError, Book> = TODO()
+  }
+}
+```
+<!--- KNIT example-validation-04.kt -->
+
+Let's forget about validating each author for a moment, and just implement the
+emptiness checks for the title and the authors list.
+
+<!--- INCLUDE
+sealed interface BookValidationError
+object EmptyTitle: BookValidationError
+object NoAuthors: BookValidationError
+data class EmptyAuthor(val index: Int): BookValidationError
+
+data class Author(val name: String)
+-->
+
+```kotlin
+data class Book private constructor(val title: String, val authors: List<Author>) {
+  companion object {
+    operator fun invoke(
+      title: String, authors: List<String>
+    ): Either<BookValidationError, Book> = either {
+      ensure(title.isNotEmpty()) { EmptyTitle }
+      ensure(authors.isNotEmpty()) { NoAuthors }
+      Book(title, TODO())
+    }
+  }
+}
+```
+<!--- KNIT example-validation-05.kt -->
+
+This code has a problem, though: it only returns _one_ error, even if there are
+two problems with the data of the `Book`. We would rather use an _accumulation_
+approach, so we can give back as much information as possible to the user.
+This requires two changes to the code:
+- The result type is now a `NonEmptyList` of problems,
+- We need to wrap the different validations in `zipOrAccumulate`.
+
+<!--- INCLUDE
+sealed interface BookValidationError
+object EmptyTitle: BookValidationError
+object NoAuthors: BookValidationError
+data class EmptyAuthor(val index: Int): BookValidationError
+
+data class Author(val name: String)
+-->
+
+```kotlin
+data class Book private constructor(val title: String, val authors: List<Author>) {
+  companion object {
+    operator fun invoke(
+      title: String, authors: List<String>
+    ): Either<NonEmptyList<BookValidationError>, Book> = either {
+      zipOrAccumulate(
+        { ensure(title.isNotEmpty()) { EmptyTitle } },
+        { ensure(authors.isNotEmpty()) { NoAuthors } }
+      ) { _, _ ->
+        Book(title, TODO())
+      }
+    }
+  }
+}
+```
+<!--- KNIT example-validation-06.kt -->
+
+The result of each of the arguments of `zipOrAccumulate` are available in the
+trailing lambda. In this case we are not using them: we have the `title`
+already available, and for the list of authors we still need to perform the
+conversion from `String` to `Author`.
+
+:::tip Arguments to zipOrAccumulate
+
+All but the last argument to `zipOrAccumulate` represent the different
+validations we want to run aggregating their output. Those arguments are
+often _blocks_ wrapped in `{` curly braces `}`, which is a bit unusual for
+most Kotliners.
```

Comment:
> which is a bit unusual for most Kotliners.
 
🤔 Is it? I was hoping it wouldn't be unusual 😅

Reply:
I got from the workshop that most people are not used to it.

Reply:
Damn, that's too bad. Should we include a section on `Either.zipOrAccumulate` which works _value based_?